src/llama.cpp

```diff
@@ -7382,6 +7449,19 @@ static struct ggml_tensor * llm_build_ffn(
                 cur = ggml_sqr(ctx, cur);
                 cb(cur, "ffn_sqr(relu)", il);
             } break;
+        case LLM_FFN_SWIGLU:
+            {
+                // Project to 4h. If using swiglu double the output width, see https://arxiv.org/pdf/2002.05202.pdf
+                int64_t split_point = cur->ne[0] / 2;
+                struct ggml_tensor * x0 = ggml_cont(ctx, ggml_view_2d(ctx, cur, split_point, cur->ne[1], cur->nb[1], 0));
+                struct ggml_tensor * x1 = ggml_cont(ctx, ggml_view_2d(ctx, cur, split_point, cur->ne[1], cur->nb[1], split_point * ggml_element_size(cur)));
+
+                x0 = ggml_silu(ctx, x0);
+                cb(cur, "ffn_silu", il);
+
+                cur = ggml_mul(ctx, x0, x1);
+                cb(cur, "ffn_mul", il);
+            } break;
```

Comment:
Instead of introducing this new activation path, you can use `LLM_FFN_SILU` + `LLM_FFN_PAR`. Just need to split the `ffn_up` tensor into `ffn_up` + `ffn_gate` during conversion

Reply:
The order of the split tensor running formulas is reversed. Assume that the split tensors are `x0` and `x1`. The original implementation is `silu(x0) * x1`. Now if `x1` is assigned to gate, the executed formula is `silu(x0 * x1)`. So it should still be necessary to add an implementation of swigrelu.

Reply:
It's not reversed, `LLM_FFN_PAR` computes exactly `silu(x0) * x1`. But I just noticed that we already have the same code for Phi3, so in that case we can introduce `LLM_FFN_SWIGLU` and reuse it in both models